subxt/src/events.rs

```diff
@@ -38,17 +37,357 @@ use codec::{
     Input,
 };
 use derivative::Derivative;
+use futures::{
+    Future,
+    FutureExt,
+    Stream,
+    StreamExt,
+};
+use jsonrpsee::core::client::Subscription;
 use scale_info::{
     PortableRegistry,
     TypeDef,
     TypeDefPrimitive,
 };
-use sp_core::Bytes;
+use sp_core::{
+    storage::StorageKey,
+    twox_128,
+    Bytes,
+};
+use std::{
+    marker::Unpin,
+    task::Poll,
+};
+
+/// Obtain events at some block hash. The generic parameter is what we
+/// will attempt to decode each event into if using [`Events::iter()`],
+/// and is expected to be the outermost event enum that contains all of
+/// the possible events across all pallets.
+///
+/// **Note:** This function is hidden from the documentation
+/// and is exposed only to be called via the codegen. Thus, prefer to use
+/// `api.events().at(block_hash)` over calling this directly.
+#[doc(hidden)]
+pub async fn at<T: Config, Evs: Decode>(
+    client: &'_ Client<T>,
+    block_hash: T::Hash,
+) -> Result<Events<'_, T, Evs>, BasicError> {
+    let mut event_bytes = client
+        .rpc()
+        .storage(&system_events_key(), Some(block_hash))
+        .await?
+        .map(|s| s.0)
+        .unwrap_or_else(Vec::new);
+
+    // event_bytes is a SCALE encoded vector of events. So, pluck the
+    // compact encoded length from the front, leaving the remaining bytes
+    // for our iterating to decode.
+    //
+    // Note: if we get no bytes back, avoid an error reading vec length
+    // and default to 0 events.
+    let cursor = &mut &*event_bytes;
+    let num_events = <Compact<u32>>::decode(cursor).unwrap_or(Compact(0)).0;
+    let event_bytes_len = event_bytes.len();
+    let remaining_len = cursor.len();
+    event_bytes.drain(0..event_bytes_len - remaining_len);
+
+    Ok(Events {
```

Comment:
Arguably this method could be a inherent constructor on `Events`, though not as important since it is used only by the macro.

Reply:
I went back and forth on this a bit, but it needed to duplicate some type params etc anyway so it felt cleaner ultimately to keep it separate :)